File: backend/tutan_agent/core/ref_system.py
from typing import List, Dict, Any, Optional
from dataclasses import dataclass

@dataclass
class RefNode:
    ref_id: str
    role: str
    text: str
    content_description: str
    resource_id: str
    bounds: Dict[str, int]
    clickable: bool
    editable: bool

class RefSystem:
    """
    Semantic Reference System inspired by OpenClaw.
    Assigns stable IDs (e1, e2...) to UI elements for LLM reasoning.
    """
    def __init__(self):
        self.current_refs: Dict[str, RefNode] = {}
        self._ref_counter = 0

    def reset(self):
        self.current_refs = {}
        self._ref_counter = 0

    def parse_aria_tree(self, tree_json: Dict[str, Any]) -> str:
        """
        Convert raw Aria Tree JSON to a flattened reference map and return a text representation for LLM.
        """
        self.reset()
        flattened_nodes = []
        self._traverse(tree_json, flattened_nodes)
        
        # Build text representation
        lines = []
        for node in flattened_nodes:
            # Only include potentially interactive or informative nodes
            if node.text or node.content_description or node.clickable or node.editable:
                self.current_refs[node.ref_id] = node
                desc = f"[{node.ref_id}] {node.role}"
                if node.text:
                    desc += f' text="{node.text}"'
                if node.content_description:
                    desc += f' label="{node.content_description}"'
                if node.clickable:
                    desc += " [clickable]"
                if node.editable:
                    desc += " [editable]"
                lines.append(desc)
        
        return "\n".join(lines)

    def _traverse(self, node: Dict[str, Any], result: List[RefNode]):
        self._ref_counter += 1
        ref_id = f"e{self._ref_counter}"
        
        ref_node = RefNode(
            ref_id=ref_id,
            role=node.get("class", "unknown").split(".")[-1], # Shorten class name
            text=node.get("text", ""),
            content_description=node.get("contentDescription", ""),
            resource_id=node.get("id", ""),
            bounds=node.get("bounds", {}),
            clickable=node.get("clickable", False),
            editable=node.get("editable", False)
        )
        
        result.append(ref_node)
        
        for child in node.get("children", []):
            self._traverse(child, result)

    def get_node(self, ref_id: str) -> Optional[RefNode]:
        return self.current_refs.get(ref_id)

File: backend/tutan_agent_nextjs/src/lib/ref-system.ts
import { XMLParser } from './xml-parser';

export interface AriaNode {
  class: string;
  text: string;
  contentDescription: string;
  id: string;
  clickable: boolean;
  editable: boolean;
  bounds: {
    left: number;
    top: number;
    right: number;
    bottom: number;
  };
  children: AriaNode[];
}

export interface RefNode {
  refId: string;
  role: string;
  text: string;
  contentDescription: string;
  bounds: {
    left: number;
    top: number;
    right: number;
    bottom: number;
  };
  clickable: boolean;
  editable: boolean;
}

export class RefSystem {
  private currentRefs: Map<string, RefNode> = new Map();
  private refCounter = 0;

  reset() {
    this.currentRefs.clear();
    this.refCounter = 0;
  }

  /**
   * Parse raw XML from ADB dump and return formatted context for LLM.
   */
  parseXmlDump(xml: string): string {
    const tree = XMLParser.parse(xml);
    return this.parseAriaTree(tree);
  }

  parseAriaTree(tree: AriaNode): string {
    this.reset();
    const flattened: RefNode[] = [];
    this.traverse(tree, flattened);

    const lines: string[] = [];
    for (const node of flattened) {
      // Filter for interactive or meaningful elements
      const isInteractive = node.clickable || node.editable;
      const hasContent = node.text || node.contentDescription;
      
      if (isInteractive || hasContent) {
        this.currentRefs.set(node.refId, node);
        let desc = `[${node.refId}] ${node.role}`;
        if (node.text) desc += ` text="${node.text}"`;
        if (node.contentDescription) desc += ` label="${node.contentDescription}"`;
        if (node.clickable) desc += " [clickable]";
        if (node.editable) desc += " [editable]";
        lines.push(desc);
      }
    }
    return lines.join("\n");
  }

  private traverse(node: AriaNode, result: RefNode[]) {
    this.refCounter++;
    const refId = `e${this.refCounter}`;
    
    const role = node.class.split(".").pop() || "unknown";
    
    result.push({
      refId,
      role,
      text: node.text,
      contentDescription: node.contentDescription,
      bounds: node.bounds,
      clickable: node.clickable,
      editable: node.editable
    });

    if (node.children) {
      for (const child of node.children) {
        this.traverse(child, result);
      }
    }
  }

  getNode(refId: string): RefNode | undefined {
    return this.currentRefs.get(refId);
  }
}
